homeassistant/components/stream/hls.py

```diff
@@ -82,29 +109,151 @@ class HlsPlaylistView(StreamView):
+                )
+        if segment.complete:
+            playlist.pop()
+            playlist.extend(
+                [
+                    # Squeeze the placeholder on the same line as #EXTINF
+                    # just to keep tidy when there are no parts
+                    "{}" + f"#EXTINF:{segment.duration:.3f},",
+                    f"./segment/{segment.sequence}.m4s",
+                ]
+            )
+            # Append another line to parts because we don't include a newline before #EXTINF
+            playlist_parts.append("")
+        return (
+            "\n".join(playlist),
+            "\n".join(playlist_parts),
+            len(segment.parts_by_http_range),
+            segment.complete,
+        )
+
+    @classmethod
+    def render(cls, track: StreamOutput, ll_hls: bool) -> str:
+        """Render HLS playlist file."""
         # NUM_PLAYLIST_SEGMENTS+1 because most recent is probably not yet complete
         segments = list(track.get_segments())[-(NUM_PLAYLIST_SEGMENTS + 1) :]
 
+        # Create a copy of the last segment with the parts frozen for rendering playlist
+        segments[-1] = copy(segments[-1])
+        segments[-1].parts_by_http_range = segments[-1].parts_by_http_range.copy()
+
         # To cap the number of complete segments at NUM_PLAYLIST_SEGMENTS,
         # remove the first segment if the last segment is actually complete
         if segments[-1].complete:
             segments = segments[-NUM_PLAYLIST_SEGMENTS:]
 
+        # The track may have been updated since we froze it, but this should be good enough
+        target_duration = track.target_duration
+
         first_segment = segments[0]
         playlist = [
             "#EXTM3U",
             "#EXT-X-VERSION:6",
             "#EXT-X-INDEPENDENT-SEGMENTS",
             '#EXT-X-MAP:URI="init.mp4"',
-            f"#EXT-X-TARGETDURATION:{track.target_duration:.0f}",
+            f"#EXT-X-TARGETDURATION:{target_duration:.0f}",
             f"#EXT-X-MEDIA-SEQUENCE:{first_segment.sequence}",
             f"#EXT-X-DISCONTINUITY-SEQUENCE:{first_segment.stream_id}",
             "#EXT-X-PROGRAM-DATE-TIME:"
             + first_segment.start_time.strftime("%Y-%m-%dT%H:%M:%S.%f")[:-3]
             + "Z",
+        ]
+
+        if ll_hls:
+            part_duration = float(
+                max(
+                    (
+                        p.duration
+                        for s in segments
+                        for p in s.parts_by_http_range.values()
+                    ),
+                    default=StreamConstants.TARGET_PART_DURATION,
+                )
+            )
+            playlist.extend(
+                [
+                    f"#EXT-X-PART-INF:PART-TARGET={part_duration:.3f}",
+                    f"#EXT-X-SERVER-CONTROL:CAN-BLOCK-RELOAD=YES,PART-HOLD-BACK={2*part_duration:.3f}",
+                    f"#EXT-X-START:TIME-OFFSET=-{3*part_duration:.3f},PRECISE=YES",
```

Comment:
Want a parallel constant for EXT_X_START for LL_HLS? Matching EXT_X_START_NON_LL_HLS

Reply:
Ok